File: PRACE-MOOC/Week5/Day2/Matrix_Multiplication_OpenACC.f90
module Matrix_Multiplication_Mod  
  implicit none 
contains
  subroutine Matrix_Multiplication(a, b, c, width)
    ! Input vectors
    real(8), intent(in), dimension(:) :: a
    real(8), intent(in), dimension(:) :: b
    real(8), intent(out), dimension(:) :: c
    real(8) :: sum = 0
    integer :: i, row, col, width, length 
    length = width*width
    !$acc kernels copyin(a(1:length), b(1:length)) copyout(c(1:length))
    !$acc loop collapse(3) reduction(+:sum)
    do row = 0, width-1
       do col = 0, width-1        
          do i = 0, width-1
             sum = sum + (a((row*width)+i+1) * b((i*width)+col+1))
          enddo
          c(row*width+col+1) = sum
          sum = 0
       enddo
    end do
    !$acc end kernels
    
  end subroutine Matrix_Multiplication
end module Matrix_Multiplication_Mod

program main
  use openacc
  use Matrix_Multiplication_Mod
  implicit none
  ! Input vectors
  real(8), dimension(:), allocatable :: a
  real(8), dimension(:), allocatable :: b 
  ! Output vector
  real(8), dimension(:), allocatable :: c
 ! real(8) :: sum = 0

  integer :: n, i 
  print *, "This program does the addition of two vectors "
  print *, "Please specify the vector size = "
  read *, n

  ! Allocate memory for vector
  allocate(a(n*n))
  allocate(b(n*n))
  allocate(c(n*n))
  
  ! Initialize content of input vectors, 
  ! vector a[i] = sin(i)^2 vector b[i] = cos(i)^2
  do i = 1, n*n
     a(i) = 1 !!sin(i*1D0) * sin(i*1D0)
     b(i) = 2 !!cos(i*1D0) * cos(i*1D0) 
  enddo
    
  ! Call the vector add subroutine 
  call Matrix_Multiplication(a, b, c, n)

  !!Verification
  do i=1,n*n
     print *, c(i)
  enddo
  
  ! Delete the memory
  deallocate(a)
  deallocate(b)
  deallocate(c)
  
end program main

File: PRACE-MOOC/Week5/Day2/Matrix_Multiplication_OpenACC.c
//-*-C++-*-
#include<stdio.h>
#include<stdlib.h>

void Matrix_Multiplication(float *a, float *b, float *restrict c, int width)   
{ 
  int length = width*width;
  float sum = 0;
#pragma acc kernels copyin(a[0:(length)], b[0:(length)]) copyout(c[0:(length)]) 
#pragma acc loop collapse(3) reduction (+:sum) 
  for(int row = 0; row < width ; ++row)                           
    {                                                             
      for(int col = 0; col < width ; ++col)                       
	{                                                         
	  for(int i = 0; i < width ; ++i)                         
	    {                                                     
	      sum += a[row*width+i] * b[i*width+col];      
	    }                                                     
	  c[row*width+col] = sum;
	  sum = 0;                            
	}
    }
}

int main()
{  
  printf("Programme assumes that matrix size is N*N \n");
  printf("Please enter the N size number \n");
  int N =0;
  scanf("%d", &N);

  // Initialize the memory on the host
  float *a, *b, *c;       
    
  // Allocate host memory
  a = (float*)malloc(sizeof(float) * (N*N));
  b = (float*)malloc(sizeof(float) * (N*N));
  c = (float*)malloc(sizeof(float) * (N*N));
  
  // Initialize host arrays
  for(int i = 0; i < (N*N); i++)
    {
      a[i] = 1.0f;
      b[i] = 2.0f;
    }
   
  // Device fuction call 
  Matrix_Multiplication(a, b, c, N);
  
  // Verification
  for(int i = 0; i < N; i++)
    {
      for(int j = 0; j < N; j++)
      	{
	  printf("%f ", c[j]);

	}
      printf("\n");
    }
  
  // Deallocate host memory
  free(a); 
  free(b); 
  free(c);

  return 0;
}
